File: P2Pclipboard/P2Pclipboard/MessageProtocol.swift
import Foundation


// MARK: - Transport Type
enum TransportType {
    case ble  // Bluetooth Low Energy - requires chunking
    case tcp  // TCP - can send as single message
}

// MARK: - Protocol Handler
class MessageProtocol {
    // BLE packet size constraint
    private static let bleMaxChunkSize = 519 - headerSize
    
    // Current protocol version
    private static let protocolVersion: UInt16 = 1
    
    // Header size: 4 (length) + 2 (version) + 1 (type) + 4 (transferId) + 4 (chunkIndex) + 4 (totalChunks)
    private static let headerSize = 19 // Increased from 13 to 19 due to length field change + total chunks
    
    private static var nextTransferId: UInt32 = 0
    
    // MARK: - Encoding
    
    private static func generateTransferId() -> UInt32 {
        let id = nextTransferId
        nextTransferId = nextTransferId &+ 1
        return id
    }

    
    static func encodeMessage(contentType: MessageContentType, payload: Data, transport: TransportType) -> [Data] {
        // Generate a unique transfer ID for this message
        let transferId = generateTransferId()
        
        // Encrypt the payload if encryption is set up
        guard let encryptedPayload = ClipboardEncryption.encrypt(data: payload) else {
            print("Failed to encrypt payload or encryption not configured")
            return [] // Return empty array to indicate failure
        }
        
        switch transport {
        case .tcp:
            // For TCP, send as one chunk regardless of size
            var chunk = Data()
            
            // Calculate total length of this message (header + payload)
            let totalLength: UInt32 = UInt32(headerSize + encryptedPayload.count)
            
            // Append headers
            chunk.append(contentsOf: totalLength.bigEndianBytes)   // Now 4 bytes
            chunk.append(contentsOf: protocolVersion.bigEndianBytes)
            chunk.append(contentType.rawValue)
            chunk.append(contentsOf: transferId.bigEndianBytes)
            chunk.append(contentsOf: UInt32(0).bigEndianBytes)     // Chunk index 0
            chunk.append(contentsOf: UInt32(1).bigEndianBytes)     // Total chunks 1
            
            // Append payload
            chunk.append(encryptedPayload)
        
            return [chunk]
            
        case .ble:
            // For BLE, we need to chunk the data
            let chunks = chunkedData(data: encryptedPayload, chunkSize: bleMaxChunkSize)
            let totalChunks = chunks.count
            
            // Create a data chunk for each piece
            return chunks.enumerated().map { index, chunkData in
                var chunk = Data()
                
                // Calculate total length of this chunk (header + payload)
                let totalLength: UInt32 = UInt32(headerSize + chunkData.count)
                
                // Append headers
                chunk.append(contentsOf: totalLength.bigEndianBytes) //totalLength
                chunk.append(contentsOf: protocolVersion.bigEndianBytes) //protocolVersion
                chunk.append(contentType.rawValue) // Type
                chunk.append(contentsOf: transferId.bigEndianBytes) //transferID
                chunk.append(contentsOf: UInt32(index).bigEndianBytes) //number of this chunk
                chunk.append(contentsOf: UInt32(totalChunks).bigEndianBytes)//total chunks
                
                // Append payload
                chunk.append(chunkData)
                
                return chunk
            }
        }
    }
    
    /// Convenience method for encoding text messages
    static func encodeTextMessage(text: String, transport: TransportType) -> [Data] {
        guard let data = text.data(using: .utf8) else {
            return []
        }
        
        return encodeMessage(contentType: .plainText, payload: data, transport: transport)
    }
    
    // MARK: - Decoding
    
    /// Message structure returned by the decoder
    struct Message {
        let contentType: MessageContentType
        let transferId: UInt32
        let payload: Data
        
        var stringPayload: String? {
            guard contentType == .plainText || contentType == .htmlContent else {
                return nil
            }
            
            print("trying to return string form stringPayload method")
            return String(data: payload, encoding: .utf8)
        }
    }
    
    /// Message chunk structure used internally for reassembly
    private struct MessageChunk {
        let contentType: MessageContentType
        let transferId: UInt32
        let chunkIndex: Int
        let totalChunks: Int
        let payload: Data
    }
    
    /// In-memory store of partial messages being reassembled
    private static var partialMessages: [UInt32: [MessageChunk]] = [:]
    
    /// Processes a received data packet according to the protocol
    /// - Parameter data: Raw received data
    /// - Returns: Complete message if available, nil if more chunks are expected
    static func decodeData(_ data: Data) -> Message? {
        // Ensure we have at least a complete header
        guard data.count >= headerSize else {
            print("Received data too small for a valid header")
            return nil
        }
        
        // Extract header fields
        let length = UInt32(bigEndianBytes: [data[0], data[1], data[2], data[3]])
        let version = UInt16(bigEndianBytes: [data[4], data[5]])
        let typeRaw = data[6]
        let transferId = UInt32(bigEndianBytes: [data[7], data[8], data[9], data[10]])
        let chunkIndex = UInt32(bigEndianBytes: [data[11], data[12], data[13], data[14]])
        let totalChunks = UInt32(bigEndianBytes: [data[15], data[16], data[17], data[18]])
        
        // Validate message type
        guard let contentType = MessageContentType(rawValue: typeRaw) else {
            print("Unknown content type: \(typeRaw)")
            return nil
        }
        
        // Extract payload
        let payload = data.suffix(from: headerSize)
        
        // Create chunk
        let chunk = MessageChunk(
            contentType: contentType,
            transferId: transferId,
            chunkIndex: Int(chunkIndex),
            totalChunks: Int(totalChunks),
            payload: payload
        )
        
        // For single-chunk messages, return immediately
        if totalChunks == 1 {
            
            guard let decryptedPayload = ClipboardEncryption.decrypt(encryptedData: payload) else {
                print("Failed to decrypt the payload with secret key")
                return nil// Return empty array to indicate failure
            }
            
            let message:Message = Message(contentType: contentType, transferId: transferId, payload: decryptedPayload)
            print("We've just decoded a complete message: \(message)")
            return message
        }
        
        // Store this chunk for multi-part message
        if partialMessages[transferId] == nil {
            partialMessages[transferId] = []
        }
        partialMessages[transferId]?.append(chunk)
        
        // Check if we have all chunks for this message
        if let chunks = partialMessages[transferId], chunks.count == Int(totalChunks) {
            // Sort chunks by index
            let sortedChunks = chunks.sorted { $0.chunkIndex < $1.chunkIndex }
            
            // Combine payloads
            var fullPayload = Data()
            for chunk in sortedChunks {
                fullPayload.append(chunk.payload)
            }
            
            // Clean up the temporary storage
            partialMessages.removeValue(forKey: transferId)
            
            guard let decryptedPayload = ClipboardEncryption.decrypt(encryptedData: fullPayload) else {
                print("Failed to decrypt the payload with secret key")
                return nil// Return empty array to indicate failure
            }
            
            print("Ok, we fully reassembled the message packed into \(Int(totalChunks)) chunks")
            
            // Return the reassembled message
            return Message(contentType: contentType, transferId: transferId, payload: decryptedPayload)
        }
        
        print("Waiting for more chunks , \(partialMessages[transferId]?.count ?? 0)/\(Int(totalChunks))")
        
        // Still waiting for more chunks
        return nil
    }
    
    // MARK: - Helper Methods
    
    /// Splits data into chunks of specified size, respecting UTF-8 character boundaries
    private static func chunkedData(data: Data, chunkSize: Int) -> [Data] {
        var chunks: [Data] = []
        var position = 0
        
        while position < data.count {
            // Calculate the end position for this chunk
            var endPos = min(position + chunkSize, data.count)
            
            // If we're not at the end of the data and the end position might be in the middle of a UTF-8 character
            if endPos < data.count {
                // Check if we're in the middle of a UTF-8 multi-byte character
                // UTF-8 continuation bytes always start with bits 10xxxxxx (so value & 0xC0 == 0x80)
                while endPos > position && (data[endPos] & 0xC0) == 0x80 {
                    // Move back to find the start of the character
                    endPos -= 1
                }
            }
            
            // Extract the chunk from position to endPos
            let chunk = data[position..<endPos]
            chunks.append(chunk)
            
            // Move to the next position
            position = endPos
        }
        
        return chunks
    }
}

File: Windows/src/MessageProtocol.cpp
#include "MessageProtocol.h"
#include "ByteUtils.h"
#include "ClipboardEncryption.h"
#include <chrono>
#include <algorithm>
#include <string>
#include <iostream>

// Initialize static members
std::map<uint32_t, std::vector<MessageProtocol::MessageChunk>> MessageProtocol::partialMessages;
std::map<uint32_t, uint64_t> MessageProtocol::partialMessageTimestamps;
uint32_t MessageProtocol::nextTransferId = 0;

std::string MessageProtocol::Message::getStringPayload() const {
    if (contentType != MessageContentType::PLAIN_TEXT &&
        contentType != MessageContentType::HTML_CONTENT) {
        return "";
    }

    return std::string(payload.begin(), payload.end());
}
// Helper method to get binary payload
const std::vector<uint8_t>& MessageProtocol::Message::getBinaryPayload() const {
    return payload;
}


uint32_t MessageProtocol::generateTransferId() {
    uint32_t id = nextTransferId;
    nextTransferId++;
    return id;
}

std::vector<std::vector<uint8_t>> MessageProtocol::encodeMessage(
    MessageContentType contentType,
    const std::vector<uint8_t>& payload,
    TransportType transport
) {
    // Generate a unique transfer ID for this message
    uint32_t transferId = generateTransferId();

    // Encrypt the payload
    std::vector<uint8_t> encryptedPayload = ClipboardEncryption::encrypt(payload);
    if (encryptedPayload.empty()) {
        std::cerr << "Failed to encrypt payload or encryption not configured" << std::endl;
        return {}; // Return empty vector to indicate failure
    }

    if (transport == TransportType::TCP) {
        // For TCP, send as one chunk regardless of size
        std::vector<uint8_t> chunk;

        // Calculate total length of this message (header + payload)
        uint32_t totalLength = HEADER_SIZE + static_cast<uint32_t>(payload.size());

        // Reserve space for the entire message
        chunk.reserve(totalLength);

        // Append headers
        auto lengthBytes = ByteUtils::uint32ToBytes(totalLength);
        chunk.insert(chunk.end(), lengthBytes.begin(), lengthBytes.end());

        auto versionBytes = ByteUtils::uint16ToBytes(PROTOCOL_VERSION);
        chunk.insert(chunk.end(), versionBytes.begin(), versionBytes.end());

        chunk.push_back(static_cast<uint8_t>(contentType));

        auto transferIdBytes = ByteUtils::uint32ToBytes(transferId);
        chunk.insert(chunk.end(), transferIdBytes.begin(), transferIdBytes.end());

        auto chunkIndexBytes = ByteUtils::uint32ToBytes(0);  // Chunk index 0 (now 4 bytes)
        chunk.insert(chunk.end(), chunkIndexBytes.begin(), chunkIndexBytes.end());

        auto totalChunksBytes = ByteUtils::uint32ToBytes(1);  // Total chunks 1 (now 4 bytes)
        chunk.insert(chunk.end(), totalChunksBytes.begin(), totalChunksBytes.end());

        // Append payload
        chunk.insert(chunk.end(), payload.begin(), payload.end());

        return { chunk };
    }
    else {
        // For BLE, we need to chunk the data
        auto chunks = chunkedData(payload, BLE_MAX_CHUNK_SIZE);
        uint32_t totalChunks = static_cast<uint32_t>(chunks.size());  // Now 4-byte value

        // Create a data chunk for each piece
        std::vector<std::vector<uint8_t>> encodedChunks;
        encodedChunks.reserve(totalChunks);

        for (uint32_t index = 0; index < totalChunks; index++) {
            std::vector<uint8_t> chunk;

            // Calculate total length of this chunk (header + payload)
            uint32_t chunkLength = HEADER_SIZE + static_cast<uint32_t>(chunks[index].size());
            chunk.reserve(chunkLength);

            // Append headers
            auto lengthBytes = ByteUtils::uint32ToBytes(chunkLength);
            chunk.insert(chunk.end(), lengthBytes.begin(), lengthBytes.end());

            auto versionBytes = ByteUtils::uint16ToBytes(PROTOCOL_VERSION);
            chunk.insert(chunk.end(), versionBytes.begin(), versionBytes.end());

            chunk.push_back(static_cast<uint8_t>(contentType));

            auto transferIdBytes = ByteUtils::uint32ToBytes(transferId);
            chunk.insert(chunk.end(), transferIdBytes.begin(), transferIdBytes.end());

            auto chunkIndexBytes = ByteUtils::uint32ToBytes(index);  // Now 4 bytes
            chunk.insert(chunk.end(), chunkIndexBytes.begin(), chunkIndexBytes.end());

            auto totalChunksBytes = ByteUtils::uint32ToBytes(totalChunks);  // Now 4 bytes
            chunk.insert(chunk.end(), totalChunksBytes.begin(), totalChunksBytes.end());

            // Append payload
            chunk.insert(chunk.end(), chunks[index].begin(), chunks[index].end());

            encodedChunks.push_back(std::move(chunk));
        }

        return encodedChunks;
    }
}

std::vector<std::vector<uint8_t>> MessageProtocol::encodeTextMessage(
    const std::string& text,
    TransportType transport
) {
    std::vector<uint8_t> payload(text.begin(), text.end());
    return encodeMessage(MessageContentType::PLAIN_TEXT, payload, transport);
}

std::shared_ptr<MessageProtocol::Message> MessageProtocol::decodeData(
    const std::vector<uint8_t>& data
) {
    std::cout << "[decodeData] Received data of size: " << data.size() << std::endl;

    if (data.size() < HEADER_SIZE) {
        std::cout << "[decodeData] Data too small for header. Returning nullptr." << std::endl;
        return nullptr;
    }

    uint32_t length = ByteUtils::bytesToUint32(data, 0);
    uint16_t version = ByteUtils::bytesToUint16(data, 4);
    uint8_t typeRaw = data[6];
    uint32_t transferId = ByteUtils::bytesToUint32(data, 7);

    // New protocol (4-byte chunk counters)
    uint32_t chunkIndex = ByteUtils::bytesToUint32(data, 11);
    uint32_t totalChunks = ByteUtils::bytesToUint32(data, 15);

    std::cout << "[decodeData] Protocol V2 Header: length=" << length
        << " version=" << version
        << " typeRaw=" << static_cast<int>(typeRaw)
        << " transferId=" << transferId
        << " chunkIndex=" << chunkIndex
        << " totalChunks=" << totalChunks << std::endl;

    if (typeRaw < 1 || typeRaw > 6) {
        std::cout << "[decodeData] Invalid typeRaw: " << static_cast<int>(typeRaw) << ". Returning nullptr." << std::endl;
        return nullptr;
    }

    MessageContentType contentType = static_cast<MessageContentType>(typeRaw);
    std::vector<uint8_t> payload(data.begin() + HEADER_SIZE, data.end());

    if (totalChunks == 1) {
        std::cout << "[decodeData] Single-chunk message. Returning immediately." << std::endl;
        auto message = std::make_shared<Message>();
        message->contentType = contentType;
        message->transferId = transferId;
        message->payload = std::move(payload);
        // Decrypt the payload
        std::vector<uint8_t> decryptedPayload = ClipboardEncryption::decrypt(message->payload);
        if (!decryptedPayload.empty()) {
            // Replace the encrypted payload with the decrypted one
            message->payload = std::move(decryptedPayload);
        }
        else {
            std::cerr << "Failed to decrypt message payload" << std::endl;
            return nullptr;
        }
        return message;
    }

    std::cout << "[decodeData] Multi-chunk message. Storing chunk." << std::endl;

    MessageChunk chunk;
    chunk.contentType = contentType;
    chunk.transferId = transferId;
    chunk.chunkIndex = chunkIndex;
    chunk.totalChunks = totalChunks;

    // Extract payload from the correct position based on version
    if (version == 1) {
        const int HEADER_SIZE_V1 = 19;
        chunk.payload = std::vector<uint8_t>(data.begin() + HEADER_SIZE_V1, data.end());
    }
    else {
        chunk.payload = std::vector<uint8_t>(data.begin() + HEADER_SIZE, data.end());
    }

    partialMessageTimestamps[transferId] = getCurrentTimeMillis();
    partialMessages[transferId].push_back(std::move(chunk));

    std::cout << "[decodeData] Chunks received for transferId " << transferId
        << ": " << partialMessages[transferId].size() << " / " << totalChunks << std::endl;

    if (partialMessages[transferId].size() == totalChunks) {
        std::cout << "[decodeData] All chunks received. Reassembling message." << std::endl;

        std::sort(partialMessages[transferId].begin(), partialMessages[transferId].end(),
            [](const MessageChunk& a, const MessageChunk& b) {
                return a.chunkIndex < b.chunkIndex;
            });

        std::vector<uint8_t> fullPayload;
        for (const auto& chunk : partialMessages[transferId]) {
            fullPayload.insert(fullPayload.end(), chunk.payload.begin(), chunk.payload.end());
        }

        auto message = std::make_shared<Message>();
        message->contentType = contentType;
        message->transferId = transferId;
        message->payload = std::move(fullPayload);

        // Decrypt the payload
        std::vector<uint8_t> decryptedPayload = ClipboardEncryption::decrypt(message->payload);
        if (!decryptedPayload.empty()) {
            // Replace the encrypted payload with the decrypted one
            message->payload = std::move(decryptedPayload);
        }
        else {
            std::cerr << "Failed to decrypt message payload" << std::endl;
            // You could either return nullptr to indicate failure
            // or continue with the encrypted payload (not recommended)
            return nullptr;
        }

        partialMessages.erase(transferId);
        partialMessageTimestamps.erase(transferId);

        std::cout << "[decodeData] Message reassembled and returned." << std::endl;
        return message;
    }

    std::cout << "[decodeData] Waiting for more chunks." << std::endl;
    return nullptr;
}


void MessageProtocol::cleanupPartialMessages(uint64_t olderThanMilliseconds) {
    uint64_t currentTime = getCurrentTimeMillis();

    // Find IDs to remove
    std::vector<uint32_t> idsToRemove;

    for (const auto& entry : partialMessageTimestamps) {
        if (currentTime - entry.second > olderThanMilliseconds) {
            idsToRemove.push_back(entry.first);
        }
    }

    // Remove the expired partial messages
    for (uint32_t id : idsToRemove) {
        partialMessages.erase(id);
        partialMessageTimestamps.erase(id);
    }
}

std::vector<std::vector<uint8_t>> MessageProtocol::chunkedData(
    const std::vector<uint8_t>& data,
    int chunkSize
) {
    std::vector<std::vector<uint8_t>> chunks;
    size_t position = 0;

    while (position < data.size()) {
        // Calculate the end position for this chunk
        size_t endPos = (std::min)(position + static_cast<size_t>(chunkSize), data.size());

        // If we're not at the end of the data and might be in the middle of a UTF-8 character
        if (endPos < data.size()) {
            // Check if we're in the middle of a UTF-8 multi-byte character
            // UTF-8 continuation bytes always start with bits 10xxxxxx (0x80-0xBF)
            while (endPos > position && (data[endPos] & 0xC0) == 0x80) {
                // Move back to find the start of the character
                endPos--;
            }
        }

        // Create the chunk from position to endPos
        std::vector<uint8_t> chunk(data.begin() + position, data.begin() + endPos);
        chunks.push_back(std::move(chunk));

        // Move to the next position
        position = endPos;
    }

    return chunks;
}

uint64_t MessageProtocol::getCurrentTimeMillis() {
    auto now = std::chrono::system_clock::now();
    auto duration = now.time_since_epoch();
    return std::chrono::duration_cast<std::chrono::milliseconds>(duration).count();
}